### ISSUE TITLE
Support for S3 Repos or use AWS CodeCommit as the Backend

### ISSUE THREAD

**Issue by damnhandy** [takezoe/gitbucket#833]
I just realized that GitBucket is using JGit to manage the repositories. What would really improve the reliability of GitBucket is if JGit could manage data using S3. There's an interesting article here on how to use S3 as a Git repository using JGit:

http://www.fancybeans.com/blog/2012/08/24/how-to-use-s3-as-a-private-git-repository/

Another idea would be to leverage the recently released [AWS CodeCommit](http://aws.amazon.com/codecommit/). The UI for CodeCommit is rather weak and doesn't do any of the things that GitBucket does. Using CodeCommit to manage the repos on the server side would go a long way towards making the repo storage much more reliable.

**Comment by McFoggy** [comment]
If you accept to host your files on an amazon node, can't you use a SAAS solution then? Bitbucket for example is free for small teams (up to 5 I guess).

**Comment by damnhandy** [comment]
Not really. With AWS, you have the option of creating a VPC, encryption, etc., which aren't available with services like BitBucket. SaaS for BitBucket and GitHub get expensive for large groups and it makes the hosted option a better option. Then there's the issue of integrating with AD, which GitBucket does pretty easily with it's LDAP option. In an AWS environment, using Code Commit as a backend would be slightly easier from an operational perspective as handling failover with an EBS volume is kind of a pain.

**Comment by damnhandy** [comment]
Thinking about this more, and being a bit less AWS-specific, it might be better to generalize this as a feature to mirror a repo to another system. This could be any alternative Git repo such as another GitBucket instance, AWS Code Commit repo, GitHub, etc.. The functionality could be similar to what is offered by the Bitbucket Server Hook Mirror plugin:

https://github.com/englishtown/stash-hook-mirror

**Comment by jhackettpps** [comment]
I'm going to revive this issue in light of Github potentially being bought by Microsoft and say using S3 as the backend would be a really great feature. 

The way I'd envisage this being used is having the database in RDS, repositories stored in S3 or some S3 compatible application, and the Gitbucket server deployed in docker. This would make deployment and redeployment from tools like Terraform totally trivial.

**Comment by lavvy** [comment]
Yeah. This is not overly impossible. Gitbucket can already connect to rds, it already runs in docker and it can potentially use s3. I believe so because gitbucket uses jgit which can use s3 as a backend. So I think it's a matter of the community seeing the need to do that. I think it should just be a small tweak. For me I wish we should have that option.
Meanwhile did you say Microsoft is buying github

**Comment by tamsky** [comment]
> Meanwhile did you say Microsoft is buying github

true story
https://news.microsoft.com/2018/06/04/microsoft-to-acquire-github-for-7-5-billion/

**Comment by yaroot** [comment]
This is not trivial task, there are things other than bare git repos (all hard coded with assumption that it's stored in local disk).

EFS (Elastic File System) might be a better alternative and yield better performance imho.
